kpatch-build: revisit checking for fentry calls

euspectre opened the pull request:
// Github did not allow me to reopen PR #520 after I force-pushed the changes, so I opened another pull request for these.

create-diff-object now checks if the original functions have fentry calls.
If an original function to be affected by the patch does not have the
fentry call, it cannot be patched. Error is reported in that case.

kpatch_create_mcount_sections() now also takes into account if a changed
or a new function has fentry call. If it does, mcount record is
generated for it as before. If a changed or a new function has no fentry
call, it is not an error in this case.

All this fixes the following issues.

1. If an original function has no fentry call (e.g. a "notrace" function)
but the patched function has it, the original function can not be
patched, but it would only be detected when applying the patch.

2. kpatch_create_mcount_sections() crashed if a patched function had no
relocation at all.

I observed such crashes when experimenting with a modified version of
the patch "tcp_cubic: better follow cubic curve after idle period" in
CentOS 7 x64.

Besides that, for a function with the first instruction starting with
0x0f, it would be incorrectly detemined that the function had fentry call.
The first bytes of the function would be overwritten in that case.

3. create-diff-object output an error if a new (an added) function had
no fentry call. This restriction is not necessary.

Signed-off-by: Evgenii Shatokhin <eshatokhin@odin.com>

jpoimboe commented on kpatch-build/create-diff-object.c:
```diff
@@ -1377,6 +1398,19 @@ void kpatch_verify_patchability(struct kpatch_elf *kelf)
+				   sym->name);
+			errs++;
+		}
+	}
```
I think this won't work, because kpatch_verify_patchability() is called with the patched object, not the original (base) object.  The base object has the original function which will be patched.

Instead of having this check here, I think a similar check will need to be made after kpatch_compare_correlated_elements() and before kpatch_elf_teardown().  For any patched function symbols with status CHANGED, their twin should have sym->has_fentry_call set.

euspectre replied:
Ah, right.

jpoimboe commented on kpatch-build/create-diff-object.c:
```diff
@@ -2605,7 +2647,8 @@ void kpatch_create_mcount_sections(struct kpatch_elf *kelf)
 		    strcmp(rela->sym->name, "__fentry__"))
-			ERROR("function '%s' has no fentry call; unable to patch", sym->name);
+			ERROR("%s: the first relocation is not for a fentry call",
+			      sym->name);
```
I think this check can be removed now.  Or is it still possible to hit this error?

euspectre replied:
I cannot say if it is possible to hit the error there. I kept it a kind of a sanity check, like the one for 'index' below:
  if (index != nr)
		ERROR("size mismatch in funcs sections");

If new symbol structures cannot appear in the list somehow after kpatch_find_fentry_calls() is called, then, I think, the check is not necessary.

jpoimboe replied:
Ok.  I'd say let's remove it.

euspectre replied:
Yes, I am checking the updated patch now, will update this pull request if everything is OK.

euspectre commented:
Here is the v2 of the patch. I removed that excessive error check and moved the checking for fentry calls after kpatch_compare_correlated_elements(). The twins for the patched symbols should be valid then.

jpoimboe commented:
:+1:
